cmd/snap-confine/snap-confine.c

```diff
@@ -552,6 +551,10 @@ static void enter_non_classic_execution_environment(sc_invocation * inv,
 	sc_ensure_shared_snap_mount();
 	debug("unsharing snap namespace directory");
 	sc_initialize_mount_ns();
+	if (sc_feature_enabled(SC_FEATURE_REFRESH_APP_AWARENESS)) {
```

Comment:
This will break pure v2 since `sc_cgroup_mount_snapd_hierarchy()` will try to mount a v1 hierarchy. Use this instead:
```
if (sc_feature_enabled(SC_FEATURE_REFRESH_APP_AWARENESS) && !sc_cgroup_is_v2()) {
```

Alternatively, you could adjust your logic in `sc_cgroup_mount_snapd_hierarchy()` and conditionally mount cgroup2 now. Also see below about dropping SC_FEATURE_REFRESH_APP_AWARENESS completely.

Reply:
AFAICT this is fine because `sc_cgroup_is_v2()` only tells us that the /sys/fs/cgroup is a unified hierarchy. However, on a v2 system it is still possible to mount named v1 hierarchies, and we plan to use this for tracking processes on both v1 and v2 systems.

On a side note, it is still possible to disable named v1 hierarchy by passing `cgroup_no_v1=named` in the kernel command line https://elixir.bootlin.com/linux/latest/source/Documentation/admin-guide/kernel-parameters.txt#L494 We've experimented with this setting, and when disabled, mount() fails with ENOENT. At the same time, I don't suspect any distro would go as far as disabling named v1 hierarchies.

Reply:
Ah, I was thinking that on a pure v2 system it would be frowned upon to use cgroup v1 at all (and even disabled via kernel config). It doesn't seem unreasonable to use cgroup2 on a pure v2 system though. I suspect we would get a bug report on this if we didn't.

Reply:
We cannot use cgroupv2 at all. Ever.

Unlike v1 there is only one v2. Additional mounts just reveal the same hierarchy in a different place. I should revise the comments to make this explicit.

This branch contains a trick. The trick is that we can use v1 for PID tracking even on "pure v2" systems, where "pure" means "systemd only uses v2". The problem with that is we have no way to co-exist with systemd in v2 world. In v1 it was different because some controllers were unused so we could just use them (freezer) and because each mount was independent. Neither of those is true in v2 world.

Unless we find a way to co-exist with systemd (requires systemd changes IMO), cgroups grow more features (parallel v2 hierarchies) or we use something entirely different for tracking (not cgroups) I intend to use v1 name=snapd mode forever, unless it is removed from the kernel entirely.

Reply:
Note, it might be frowned upon in some sense but that's the best way out of the current situation. I think going forward we need to devise a way to have process tracking while systemd itself is managing cgroups (in a scope that, I presume, will only grow) for the needs of services.

Even non-service processes are likely to end up in a hierarchy managed by systemd via {system,user}.slice, user-1000.slice, session-2.scope, user@1000.service, and so on.

Reply:
Ok, thanks for the explanation and detailed analysis (it can be hard to keep it all straight when only thinking about pure v2 in theoretical terms). Considering all of this, we probably want to document that cgroup v1 is required in the kernel to run snapd (wherever that list of kernel config lives...). This maybe deserves a system-key tie-in as well, but that can come in a followup if we decide to do that.